gh-128508: Add some docstrings to xml.dom.minidom

srinivasreddy opened the pull request:
Initial version extracted from PR #24152, which also deleted and modified minidom tests. 

Co-authored-by: Karl Dubost [karl+github@la-grange.net](mailto:karl+github@la-grange.net)

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

<!-- gh-issue-number: gh-128508 -->
* Issue: gh-128508
<!-- /gh-issue-number -->

srinivasreddy committed:
Update and Add docstrings for functions and methods in minidom.py module

sobolevn committed:
Add `check-readthedocs` pre-commit hook (#128453)

srinivasreddy commented on Lib/xml/dom/minidom.py:
```diff
@@ -882,7 +898,7 @@ def getElementsByTagNameNS(self, namespaceURI, localName):
 
     def __repr__(self):
-        return "<DOM Element: %s at %#x>" % (self.tagName, id(self))
+        return f"<DOM Element: {self.tagName}>"
```
This breaks the backward compatibility. But should we care about the programs that depend on `__repr__` ?
Isn't that implementation detail?

terryjreedy replied:
This is off-topic for a docstring patch, which I believe can otherwise be backported.  Please delete.  (In case you don't know, please do not squash commits.  )

srinivasreddy replied:
Addressed in  b0a91293f00

erlend-aasland committed:
gh-128152: Argument Clinic: ignore pre-processor directives inside C comments (#128464)

barneygale committed:
GH-127381: pathlib ABCs: remove `PathBase.move()` and `move_into()` (#128337)

These methods combine `_delete()` and `copy()`, but `_delete()` isn't part
of the public interface, and it's unlikely to be added until the pathlib
ABCs are made official, or perhaps even later.

erlend-aasland committed:
Docs: mark up json.dump() using parameter list (#128482)

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

barneygale committed:
pathlib tests: create `walk()` test hierarchy without using class under test (#128338)

In the tests for `pathlib.Path.walk()`, avoid using the path class under
test (`self.cls`) in test setup. Instead we use `os` functions in
`test_pathlib`, and direct manipulation of `DummyPath` internal data in
`test_pathlib_abc`.

StanFromIreland committed:
gh-126719: Clarify math.fmod docs (#127741)

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

erlend-aasland committed:
Docs: amend json.dump() post gh-128482 (#128489)

rruuaanng committed:
gh-127954: Document PyObject_DelItemString (#127986)

bombs-kim committed:
gh-127553: Remove outdated TODO comment in _pydatetime (#127564)

erlend-aasland committed:
gh-115765: Document and enforce Autoconf 2.72 requirement (#128502)

terryjreedy commented on Lib/xml/dom/minidom.py:
```diff
@@ -80,10 +82,16 @@ def _get_lastChild(self):
             return self.childNodes[-1]
 
     def insertBefore(self, newChild, refChild):
+        """Insert a new DOM Node before an existing Node.
```
Blank line is needed.
```suggestion
        """Insert a new DOM Node before an existing Node.

```
I think first url should be just above 2nd.

terryjreedy commented on Lib/xml/dom/minidom.py:
```diff
@@ -172,11 +183,15 @@ def removeChild(self, oldChild):
         return oldChild
 
     def normalize(self):
+        """Transform a node into its normalized form.
```
```suggestion
        """Transform a node into its normalized form.

```

terryjreedy commented on Lib/xml/dom/minidom.py:
```diff
@@ -1904,7 +1921,8 @@ def renameNode(self, n, namespaceURI, name):
 def _clone_node(node, deep, newOwnerDocument):
     """
-    Clone a node and give it the new owner document.
+    Returns a copy of node.
```
Remove \n after """.  Add after this line.  But this revision seems wrong as it omits mention of newOwnerDocument.  I would have to read code to suggest improvement.

terryjreedy commented on Lib/xml/dom/minidom.py:
```diff
@@ -204,6 +219,7 @@ def normalize(self):
         self.childNodes[:] = L
 
     def cloneNode(self, deep):
+        """The Node.cloneNode() method returns a duplicate of the node."""
```
Should be "Return a duplicate ...", but needs to explain 'deep'  (I suspect it is a boolean.)  Docstrings are for quick reminders when coding, and 'deep' is not obvious.

terryjreedy commented on Lib/xml/dom/minidom.py:
```diff
@@ -1341,6 +1357,7 @@ def _get_internalSubset(self):
         return self.internalSubset
 
     def cloneNode(self, deep):
+        """The Node.cloneNode() method returns a duplicate of the node."""
```
See other cloneNode comment.

terryjreedy commented on Lib/xml/dom/minidom.py:
```diff
@@ -129,6 +138,7 @@ def appendChild(self, node):
         return node
 
     def replaceChild(self, newChild, oldChild):
+        """Replace an existing node with a new node."""
```
```suggestion
        """Replace child node *oldChild* with *newChild*."""
```

terryjreedy reviewed (changes requested):
I have not worked on docstrings, especially not for anything other that IDLE, for years.  The best reference I know is the Multiline String section of https://peps.python.org/pep-0257/, though the thoroughness described there may be rare.  Still, these need more work.  The xml.dom and xml.dom.minidom docs should be a basis for corresponding docstrings.  Usefully summarizing behavior can be hard.

I think a blurb should be added, but this can wait.

zanieb committed:
gh-128437: Add `BOLT_COMMON_FLAGS` with `-update-debug-sections` (gh-128455)

Add `BOLT_COMMON_FLAGS` with `-update-debug-sections`

Co-authored-by: Gregory Szorc <gregory.szorc@gmail.com>

corona10 committed:
gh-128137: Update PyASCIIObject to handle interned field with the atomic operation (gh-128196)

Damien-Chen committed:
gh-128504: Upgrade doctest to ubuntu-24.04 (#128506)

Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>
Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

Commit:
Docs: fix `MessageDefect` references in email.policy docs (#128468)

RanKKI committed:
gh-98188: Fix EmailMessage.get_payload to decode data when CTE value has extra text (#127547)

Up to this point message handling has been very strict with regards to content encoding values: mixed case was accepted, but trailing blanks or other text would cause decoding failure, even if the first token was a valid encoding.  By Postel's Rule we should go ahead and decode as long as we can recognize that first token.  We have not thought of any security or backward compatibility concerns with this fix.

This fix does introduce a new technique/pattern to the Message code: we look to see if the header has a 'cte' attribute, and if so we use that.  This effectively promotes the header API exposed by HeaderRegistry to an API that any header parser "should" support.  This seems like a reasonable thing to do.  It is not, however, a requirement, as the string value of the header is still used if there is no cte attribute.

The full fix (ignore any trailing blanks or blank-separated trailing text) applies only to the non-compat32 API.  compat32 is only fixed to the extent that it now ignores trailing spaces.  Note that the HeaderRegistry parsing still records a HeaderDefect if there is extra text.

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

srinivasreddy committed:
Update Lib/xml/dom/minidom.py

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

srinivasreddy committed:
Update Lib/xml/dom/minidom.py

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

srinivasreddy committed:
Update Lib/xml/dom/minidom.py

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

srinivasreddy committed:
Revert __repr__ change

srinivasreddy committed:
Merge branch 'main' into gh-63882-doc_strings

srinivasreddy committed:
Update the docstring for _clone_node(...)

srinivasreddy committed:
Update docstring for cloneNode(...)

srinivasreddy committed:
Update docstrings for cloneNode(...)

erlend-aasland commented on Lib/xml/dom/minidom.py:
```diff
@@ -172,11 +184,16 @@ def removeChild(self, oldChild):
     def normalize(self):
+        """Transform a node into its normalized form.
+
+        Removes empty exclusive Text nodes and concatenates the data of
```
Please use the imperative mode for reference text (docstrings count as reference).

srinivasreddy replied:
Addressed in a4840f33d3d3c3292663a53e3a205feb1b8e3210

erlend-aasland commented on Lib/xml/dom/minidom.py:
```diff
@@ -204,6 +221,12 @@ def normalize(self):
+
+        Args:
+            deep (bool): If True, recursively clone this node's descendants.
+                        If False, clone only this node.
```
There is no _"Args:"_ convention for parameter docstrings. Please take a look at how Argument Clinic generates parameters docstrings and follow that convention.

erlend-aasland replied:
FTR, 727af86547291334a16a44ccd6b7c8c9f31f20e0 does _not_ align with Argument Clinic's formatting. Please take another look at how it is done.

srinivasreddy replied:
You mean this way?

```
def cloneNode(self, deep):
        """Create and return a duplicate of this node.

        :param deep: If True, recursively clone this node's descendants.
                     If False, clone only this node.
        """
        return _clone_node(self, deep, self.ownerDocument or self)
```

erlend-aasland replied:
No. Please take a look at how Argument Clinic generates docstrings. You'll find Argument Clinic generated code in the `clinic/` directories in the repo. For example:

https://github.com/python/cpython/blob/8d15058d61681e98579cf3fbd3c6fd13b3df6a72/Modules/_sqlite/clinic/connection.c.h#L233-L248

```sh
$ python3 -c "import sqlite3; print(sqlite3.Connection.blobopen.__doc__)"
Open and return a BLOB object.

  table
    Table name.
  column
    Column name.
  row
    Row index.
  readonly
    Open the BLOB without write permissions.
  name
    Database name.
```

srinivasreddy replied:
Addressed in ecb4a54d051d24f6bb8344d92b07a4c7170c6fa3

srinivasreddy committed:
Convert doc string to imperative mode

erlend-aasland commented on Lib/xml/dom/minidom.py:
```diff
@@ -80,10 +82,17 @@ def _get_lastChild(self):
         if newChild.nodeType == self.DOCUMENT_FRAGMENT_NODE:
             for c in tuple(newChild.childNodes):
                 self.insertBefore(c, refChild)
-            ### The DOM does not clearly specify what to return in this case
```
Please revert all unrelated changes.

srinivasreddy replied:
Addressed in 8df67955262cfc6ad1556217cd2772f24c1bf40a

srinivasreddy committed:
Update docstring as recommended by argument clinic

srinivasreddy committed:
Update docstring as recommended by argument clinic

srinivasreddy committed:
Undo removing comment

srinivasreddy committed:
Update docstrings as it was done in argument clinic

srinivasreddy committed:
Add space back

srinivasreddy committed:
Update doc strings

erlend-aasland commented on Lib/xml/dom/minidom.py:
```diff
@@ -1903,9 +1934,15 @@ def renameNode(self, n, namespaceURI, name):
+    """Create and return a clone of a DOM node.
+
+      node
+        The DOM node to clone
```
```suggestion
        The DOM node to clone.
```

erlend-aasland commented on Lib/xml/dom/minidom.py:
```diff
@@ -1903,9 +1934,15 @@ def renameNode(self, n, namespaceURI, name):
+        If True, recursively clone the node's descendants.
+        If False, only clone the node itself.
+      newOwnerDocument
+        The document that will own the cloned node
```
```suggestion
        The document that will own the cloned node.
```

srinivasreddy committed:
Update Lib/xml/dom/minidom.py

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

srinivasreddy committed:
Update Lib/xml/dom/minidom.py

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

merwok commented on Lib/xml/dom/minidom.py:
```diff
@@ -32,6 +32,7 @@
 
 
 class Node(xml.dom.Node):
+    """Define properties accessible on a DOM node."""
```
I think it’s strange to have a class docstring start with a verb; it makes sense for a method (that does something), but a class usually represents a thing (noun), it does not do something.

srinivasreddy replied:
Addressed in [b97d812](https://github.com/python/cpython/pull/128477/commits/b97d812e6dde095b004d72465ed3f8837e46b96c)

merwok commented on Lib/xml/dom/minidom.py:
```diff
@@ -80,6 +82,14 @@ def _get_lastChild(self):
+        https://dom.spec.whatwg.org/#dom-node-insertbefore
+        The insertBefore(node, child) method, when invoked,
+        must return the result of pre-inserting node into
+        this before child.
```
I think the docstring should document what the method does, not quote parts of a spec.

I would suggest simpler wording, but I don’t understand the existing one («the result of pre-inserting node into this before child» ?!)

srinivasreddy replied:
Addressed in [b97d812](https://github.com/python/cpython/pull/128477/commits/b97d812e6dde095b004d72465ed3f8837e46b96c)

merwok commented on Lib/xml/dom/minidom.py:
```diff
@@ -80,6 +82,14 @@ def _get_lastChild(self):
+        The insertBefore(node, child) method, when invoked,
+        must return the result of pre-inserting node into
+        this before child.
+        See also https://dom.spec.whatwg.org/#concept-node-pre-insert
```
I would have links like this in the reST docs, not docstrings.

srinivasreddy replied:
Addressed in b97d812

srinivasreddy committed:
Merge branch 'main' into gh-63882-doc_strings

srinivasreddy committed:
Address review comments

srinivasreddy committed:
Update docstrings

erlend-aasland commented on Lib/xml/dom/minidom.py:
```diff
@@ -32,6 +32,9 @@
 
 
 class Node(xml.dom.Node):
+    """Base class representing a node in the DOM tree.
```
Missing newline:

```suggestion
    """Base class representing a node in the DOM tree.

```

srinivasreddy committed:
Update Lib/xml/dom/minidom.py

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

srinivasreddy committed:
Merge branch 'main' into gh-63882-doc_strings
